### ISSUE TITLE
Report an error if a comparison is discarded

### ISSUE THREAD

**Issue by RobertCraigie** [microsoft/pyright#3174]
**Is your feature request related to a problem? Please describe.**

Recently wrote this by accident

```py
response.status_code == 400
```

When it should have been

```py
response.status_code = 400
```

**Describe the solution you'd like**

I'd like if it pyright could report an error in this case, like it does when you accidentally discard the result of a coroutine:

```py
async def main():
    ...

async def foo() -> None:
    main()  # error result of async function call is not used.
```

**Additional context**

I understand that this may be considered the job of a linter to detect issues like this but I think it would be worth including in pyright as well.

**Comment by erictraut** [comment]
I agree this could be useful. Let's think about how this feature might be spec'ed.

I presume this would apply only in cases where an expression comprises an entire statement.

Would it apply only to `==` and `!=` operators? What about other unary, binary, and trinary operators? 

```python
not x
4 or 5
4 + 5
4 if x else 5
```

Would it apply to cases where one of the two operand expressions might have a side effect?

```python
func() == 4
```

What if the LHS operand is the instance of a class with a `__eq__` or `__ne__` method? Can we safely assume that these are not stateful and have no side effects?

```python
class Foo:
    def __eq__(self, __o: object) -> bool: ...

foo = Foo()

foo == 4
```

**Comment by RobertCraigie** [comment]
> Would it apply only to == and != operators? What about other unary, binary, and trinary operators?

I think it should apply to the unary, binary and trinary operators as well as discarding the results of those operators would also be indicative of an unintentional error.

> Would it apply to cases where one of the two operand expressions might have a side effect?

In this example the fact that the RHS is present is an error (unless the LHS returns a class that has side effects) but it is more likely a style issue instead of a programming error. For example, the original reason I requested this feature is because I intended to assign a value but accidentally discarded a comparison instead but assigning to a function call is invalid syntax:

```py
foo() = 4  # syntax error
```

Obviously function calls are not the only expressions that can have side effects so this is difficult to spec.

On the other hand if the expressions are swapped so that the side effect is on the RHS then this is highly likely to be unintentional:

```py
a == foo()
```

Even if `foo()` has side effects and the result is intended to be discarded the code is confusing and easy to read incorrectly. I would say an error should be reported if the side effect expression is on the RHS.

> What if the LHS operand is the instance of a class with a __eq__ or __ne__ method? Can we safely assume that these are not stateful and have no side effects?

I'm not sure about this one, I would think that a class that implements an `__eq__` method that has side effects would be an anti-pattern although that would not be the job of a type checker to report.

**Comment by erictraut** [comment]
I've introduced a new diagnostic rule called `reportUnusedExpression` that detects this case.

I realized that it's not safe to apply this to ternary operators, so I've limited it to binary and unary operators. 

This will be included in the next release.

**Comment by RobertCraigie** [comment]
Great thank you!

> I realized that it's not safe to apply this to ternary operators, so I've limited it to binary and unary operators.

Just for my own curiosity what makes it not safe for ternary operators?

**Comment by erictraut** [comment]
There are legitimate uses for ternary operators that have conditional side effects.

```python
do_something() if x > 1 else do_something_else()
```

**Comment by erictraut** [comment]
This is included in pyright 1.1.230, which I just published. It will also be included in the next release of pylance.
